clients/common/nm-vpn-helpers.c

```diff
@@ -247,3 +253,522 @@ nm_vpn_openconnect_authenticate_helper (const char *host,
+	nm_connection_add_setting (connection, NM_SETTING (s_con));
+	s_ip4 = NM_SETTING_IP_CONFIG (nm_setting_ip4_config_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_ip4));
+	s_ip6 = NM_SETTING_IP_CONFIG (nm_setting_ip6_config_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_ip6));
+	s_wg = NM_SETTING_WIREGUARD (nm_setting_wireguard_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_wg));
+
+	uuid = nm_utils_uuid_generate ();
+
+	g_object_set (s_con,
+	              NM_SETTING_CONNECTION_ID,
+	              ifname,
+	              NM_SETTING_CONNECTION_UUID,
+	              uuid,
+	              NM_SETTING_CONNECTION_TYPE,
+	              NM_SETTING_WIREGUARD_SETTING_NAME,
+	              NM_SETTING_CONNECTION_INTERFACE_NAME,
+	              ifname,
+	              NULL);
+
+	g_object_set (s_wg,
+	              NM_SETTING_WIREGUARD_PRIVATE_KEY,
+	              data_private_key,
+	              NM_SETTING_WIREGUARD_LISTEN_PORT,
+	              data_listen_port,
+	              NM_SETTING_WIREGUARD_FWMARK,
+	              data_fwmark,
+	              NULL);
+
+	if (data_peers) {
+		for (i = 0; i < data_peers->len; i++)
+			nm_setting_wireguard_append_peer (s_wg, data_peers->pdata[i]);
+	}
+
+	for (is_v4 = 0; is_v4 < 2; is_v4++) {
+		const char *method_disabled = is_v4 ? NM_SETTING_IP4_CONFIG_METHOD_DISABLED : NM_SETTING_IP6_CONFIG_METHOD_IGNORE;
+		const char *method_manual   = is_v4 ? NM_SETTING_IP4_CONFIG_METHOD_MANUAL   : NM_SETTING_IP6_CONFIG_METHOD_MANUAL;
+		NMSettingIPConfig *s_ip     = is_v4 ? s_ip4                                 : s_ip6;
+		GPtrArray *data_dns         = is_v4 ? data_dns_v4                           : data_dns_v6;
+		GPtrArray *data_addr        = is_v4 ? data_addr_v4                          : data_addr_v6;
+
+		if (data_dns && !data_addr) {
+			/* When specifying "DNS", we also require an "Address" for the same address
+			 * family. That is because a NMSettingIPConfig cannot have @method_disabled
+			 * and DNS settings at the same time.
+			 *
+			 * We don't have addresses. Silently ignore the DNS setting. */
+			data_dns = NULL;
+		}
+
+		g_object_set (s_ip,
+		              NM_SETTING_IP_CONFIG_METHOD,
+		              data_addr ? method_manual : method_disabled,
+		              NULL);
+
+		if (data_addr) {
+			for (i = 0; i < data_addr->len; i++)
+				nm_setting_ip_config_add_address (s_ip, data_addr->pdata[i]);
+		}
+		if (data_dns) {
+			for (i = 0; i < data_dns->len; i++)
+				nm_setting_ip_config_add_dns (s_ip, data_dns->pdata[i]);
+		}
+
+		if (data_table == _TABLE_AUTO) {
+			/* TODO: not implemented. */
+		} else if (data_table == _TABLE_OFF) {
+			/* TODO: not implemented. Well, currently we don't support adding direct routes for the
+			 * allowed-ips ranges. So, it implicitly is implemented.  */
+		} else {
+			/* TODO: setting NM_SETTING_IP_CONFIG_ROUTE_TABLE to a routing table is not exactly the same as
+			 * what wg-quick does. */
```

Comment:
What is that doing, btw?

Reply:
Let me quote the documentation (just to see whether that description is comprehensible):

```
»···/**
»··· * NMSettingIPConfig:route-table:
»··· *
»··· * Enable policy routing (source routing) and set the routing table used when adding routes.
»··· *
»··· * This affects all routes, including device-routes, IPv4LL, DHCP, SLAAC, default-routes
»··· * and static routes. But note that static routes can individually overwrite the setting
»··· * by explicitly specifying a non-zero routing table.
»··· *
»··· * If the table setting is left at zero, it is eligible to be overwritten via global
»··· * configuration. If the property is zero even after applying the global configuration
»··· * value, policy routing is disabled for the address family of this connection.
»··· *
»··· * Policy routing disabled means that NetworkManager will add all routes to the main
»··· * table (except static routes that explicitly configure a different table). Additionally,
»··· * NetworkManager will not delete any extraneous routes from tables except the main table.
»··· * This is to preserve backward compatibility for users who manage routing tables outside
»··· * of NetworkManager.
»··· *
»··· * Since: 1.10
»··· **/
```

Reply:
Wording makes sense to me, though I don't quite see how this is different from wg-quick's behavior, except the global override business.

Reply:
yeah, it's not too different. I updated the comment.